### PR TITLE
(test) O3-4553: Add unit tests for question component

### PR DESCRIPTION
**Pull request by MohamedAliMefteh**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary
This PR adds unit tests for the Question Component used within modals to ensure the correct rendering, behavior, and functionality. The tests cover a wide range of scenarios, including interactions with form fields, conditional rendering, and component-specific functionality such as converting labels to camel case, toggling question visibility, and updating question info.

However, some of the tests are currently failing and require adjustments. Additionally, a few test cases are incomplete and need to be finalized later.

### Key Changes:

* Rendering & Interactions: Tests ensure the Question component renders correctly and interacts with the form fields, such as updating the questionId, toggling visibility of additional question info, and updating question info text.

* Question Info Visibility: Tests handle toggling the visibility of the "Additional Question Info" section when radio buttons (Yes/No) are clicked.

* Label Conversion: Added tests for converting the question label to camel-case when the corresponding button is clicked.

* Component Rendering: Tests ensure that child components like RenderTypeComponent and QuestionTypeComponent are rendered when their corresponding properties are set.

### Known Issues:

* Test Failures: There are a few tests that are currently not passing. The failing tests are marked with comments in the code, and they need to be reviewed and adjusted to handle specific behaviors or edge cases correctly.

  * Duplicate Question ID Validation: The test for checking duplicate question IDs is commented out due to an issue in mocking the validation function (checkIfQuestionIdExists). It needs to be reworked to properly test the validation logic.

  * Rendering Type with Markdown: The test to ensure that label and required inputs are hidden when the rendering type is set to "markdown" is commented out. This needs to be revisited to handle conditional rendering based on the rendering property.

* Incomplete Tests: There are two tests marked with TODO comments that need to be completed later:

  * Question Required Status: The test for toggling between required and optional statuses is incomplete and needs to be updated to verify that the prop value is correctly passed down and updated.

  * Rendering Type Select Options: The test for checking the select options when rendering type is clicked is a work-in-progress and needs to be finalized to validate the available options for the user.

## Related Issue

https://openmrs.atlassian.net/browse/O3-4553

### CODE REVIEW

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+}));
+
+jest.mock('../common/required-label/required-label.component', () => ({
+  __esModule: true,
+  default: jest.fn(({ isRequired, text }) => (
+    <div data-testid="required-label">
+      {text} {isRequired ? '(Required)' : ''}
+    </div>
+  )),
+}));
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:28]: Why are these internal components mocked? We should try to avoid mocking internal components like these as much as possible.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+jest.mock('react-i18next', () => ({
+  useTranslation: jest.fn(() => ({
+    t: (key, fallback) => fallback || key,
+  })),
+}));
+
+jest.mock('./question.scss', () => ({
+  questionIdLabel: 'mock-question-id-label',
+}));
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:38]: Why is a stylesheet mocked?

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  );
+};
+
+describe('Question Component', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should render without crashing', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={jest.fn()} />);
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:66]: Define `checkIfQuestionIdExists` as a mock function at the top so you can just pass it in instead of defining it as `jest.fn` every time.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    expect(screen.getByRole('textbox', { name: /Question ID/i })).toBeInTheDocument();
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:67]: Not just the question id, the question type field, you should also check if question type and rendering type fields also get rendered.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+  // it('should validate duplicate question ids', () => {
+  //   const checkIfQuestionIdExists = jest.fn(() => true);
+  //   renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />, {
+  //     formField: { ...initialFormField, id: 'duplicateId' },
+  //   });
+
+  //   expect(checkIfQuestionIdExists).toHaveBeenCalledWith('duplicateId');
+  //   expect(screen.getByText(/This question ID already exists/i)).toBeInTheDocument();
+  // });
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:93]: Don't include commented out test cases

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    expect((questionInfoInput as HTMLInputElement).value).toBe('New question info');
+  });
+
+  it('should toggle question required status', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={jest.fn()} />);
+
+    const requiredRadio = screen.getByLabelText(/Required/i);
+    fireEvent.click(requiredRadio);
+
+    // TODO: check if it gets the updated prop
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:141]: Please don't add test cases with To Do comments.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  });
+
+  // it('should not show label and required inputs when rendering type is markdown', () => {
+  //   renderWithFormFieldProvider(<Question checkIfQuestionIdExists={jest.fn()} />, {
+  //     formField: { ...initialFormField, questionOptions: { rendering: 'markdown' } },
+  //   });
+
+  //   expect(screen.queryByRole('textbox', { name: /Label/i })).not.toBeInTheDocument();
+  //   expect(screen.queryByText(/Is this question a required or optional field?/i)).not.toBeInTheDocument();
+  // });
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:174]: Ditto

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+  });
+
+  it('should render QuestionTypeComponent when question type is selected', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={jest.fn()} />, {
+      formField: { ...initialFormField, type: 'obs' },
+    });
+
+    expect(screen.getByTestId('question-type-component')).toBeInTheDocument();
+    expect(QuestionTypeComponent).toHaveBeenCalled();
+  });
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:165]: These components should be rendered even if they are not in the field.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+  it('should show appropriate rendering types for the selected question type', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={jest.fn()} />, {
+      formField: { ...initialFormField, type: 'encounterDatetime' },
+    });
+
+    const renderingTypeSelect = screen.getByLabelText(/Rendering type/i);
+    fireEvent.click(renderingTypeSelect);
+
+    // TODO: test the select options.
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:184]: Ditto

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+  it('should toggle question required status', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={jest.fn()} />);
+
+    const requiredRadio = screen.getByLabelText(/Required/i);
+    fireEvent.click(requiredRadio);
+
+    // TODO: check if it gets the updated prop
+
+    const optionalRadio = screen.getByLabelText(/Optional/i);
+    fireEvent.click(optionalRadio);
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:144]: Prefer to use userEvents over fireEvent - https://testing-library.com/docs/user-event/intro#differences-from-fireevent

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+    // TODO: check if it gets the updated prop
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:146]: Ditto

**Review by NethmiRodrigo — CHANGES_REQUESTED** [review]
Thanks @MohamedAliMefteh, a couple of suggestions:

**Comment by MohamedAliMefteh** [comment]
> Thanks @MohamedAliMefteh, a couple of suggestions:

When writing these tests, I didn't take in consideration other tests written in openMRS form-builder.
I will try to follow the same rules within other tests.
I will tackle these changes Once I finish my proposal, I hope that's not a problem!

File: src/components/interactive-builder/modals/question/question-form/question/question.component.tsx
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.component.tsx:1]: I don't think this file should have been changed. Maybe a bad rebase?

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 12]
```diff
@@ -1,73 +1,182 @@
...
+    rendering: 'text',
+  },
 };
 
-jest.mock('../../form-field-context', () => ({
-  ...jest.requireActual('../../form-field-context'),
-  useFormField: () => ({ formField, setFormField: mockSetFormField }),
-}));
+const renderWithFormFieldProvider = (
+  ui: React.ReactElement,
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:24]:
```suggestion
  component: React.ReactElement,
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 13]
```diff
@@ -1,73 +1,182 @@
...
+  { formField = initialFormField, selectedConcept = null } = {},
+) => {
+  return render(
+    <FormFieldProvider initialFormField={formField} selectedConcept={selectedConcept}>
+      {ui}
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:29]:
```suggestion
      {component}
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 14]
```diff
@@ -1,73 +1,182 @@
...
+import { FormFieldProvider } from '../../form-field-context';
+import type { FormField } from '@openmrs/esm-form-engine-lib';
 
-const mockSetFormField = jest.fn();
-const checkIfQuestionIdExists = jest.fn();
+jest.mock('react-i18next', () => ({
+  useTranslation: jest.fn(() => ({
+    t: (key, fallback) => fallback || key,
+  })),
+}));
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:12]: Does this need to be mocked?

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 15]
```diff
@@ -1,73 +1,182 @@
...
+  const checkIfQuestionIdExists = jest.fn(() => false);
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should render all required fields', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />);
+
+    expect(screen.getByRole('textbox', { name: /Question ID/i })).toBeInTheDocument();
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:44]: Since the `/i` is there, the text doesn't need to strictly be in the case we expect.
```suggestion
    expect(screen.getByRole('textbox', { name: /question id/i })).toBeInTheDocument();
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 16]
```diff
@@ -1,73 +1,182 @@
...
+
+    expect(screen.getByLabelText(/Question type/i)).toBeInTheDocument();
+
+    expect(screen.getByLabelText(/Rendering type/i)).toBeInTheDocument();
+  });
+
+  it('should display the form field id in the question id input', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />);
+    const idInput = screen.getByRole('textbox', { name: /Question ID/i }) as HTMLInputElement;
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:53]: Ditto

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 17]
```diff
@@ -1,73 +1,182 @@
...
+    expect(idInput).toHaveValue('testId');
   });
 
-  it('shows TextInput when "Yes" is selected', async () => {
+  it('should update form field when question id changes', async () => {
     const user = userEvent.setup();
-    renderQuestionComponent();
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />);
+    const idInput = screen.getByRole('textbox', { name: /Question ID/i });
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:60]: Ditto

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 18]
```diff
@@ -1,73 +1,182 @@
...
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should render all required fields', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />);
+
+    expect(screen.getByRole('textbox', { name: /Question ID/i })).toBeInTheDocument();
+
+    expect(screen.getByLabelText(/Question type/i)).toBeInTheDocument();
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:46]:
```suggestion
    expect(screen.getByLabelText(/question type/i)).toBeInTheDocument();
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 19]
```diff
@@ -1,73 +1,182 @@
...
+
+    expect(screen.getByLabelText(/Rendering type/i)).toBeInTheDocument();
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:48]:
```suggestion
    expect(screen.getByLabelText(/rendering type/i)).toBeInTheDocument();
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 20]
```diff
@@ -1,73 +1,182 @@
...
-    expect(
-      screen.getByRole('group', { name: /would you like to provide additional details about the question?/i }),
-    ).toBeInTheDocument();
-    expect(screen.getByRole('radio', { name: /Yes/i })).toBeInTheDocument();
-    expect(screen.getByRole('radio', { name: /No/i })).toBeInTheDocument();
+  const checkIfQuestionIdExists = jest.fn(() => false);
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:39]: This is not required since this is done by default in our jest.config

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 21]
```diff
@@ -1,73 +1,182 @@
...
 describe('Question Component', () => {
-  it('should render the toggles for question info', () => {
-    renderQuestionComponent();
-
-    expect(
-      screen.getByRole('group', { name: /would you like to provide additional details about the question?/i }),
-    ).toBeInTheDocument();
-    expect(screen.getByRole('radio', { name: /Yes/i })).toBeInTheDocument();
-    expect(screen.getByRole('radio', { name: /No/i })).toBeInTheDocument();
+  const checkIfQuestionIdExists = jest.fn(() => false);
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:35]: Shouldn't this be at the top of the file?

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 22]
```diff
@@ -1,73 +1,182 @@
...
-    renderQuestionComponent();
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />);
+    const idInput = screen.getByRole('textbox', { name: /Question ID/i });
 
-    const yesRadio = screen.getByRole('radio', { name: /Yes/i });
-    await user.click(yesRadio);
+    await user.clear(idInput);
+    await user.type(idInput, 'newTestId');
+
+    expect((idInput as HTMLInputElement)).toHaveValue('newTestId');
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:65]: Does this need to be casted as an `HTMLInputElement`?

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 23]
```diff
@@ -1,73 +1,182 @@
...
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />, {
+      formField: { ...initialFormField, type: 'encounterDatetime' },
+    });
+
+    const renderingTypeSelect = screen.getByLabelText(/Rendering type/i);
+
+    expect(renderingTypeSelect).toBeInTheDocument();
+
+    expect(screen.getByLabelText(/Question type/i)).toHaveValue('encounterDatetime');
+  });
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:157]: Since this test case is supposed to test if the appropriate rendering types are shown, we should do more than checking if the select rendering type dropdown is there, we should also test the select options available. In this case, the question type is `encounterDatetime`, so the rendering type options should only be `date` or `datetime`

**Review by NethmiRodrigo — CHANGES_REQUESTED** [review]

File: src/components/interactive-builder/modals/question/question-form/question/question.component.tsx
Comment (MohamedAliMefteh) [reply src/components/interactive-builder/modals/question/question-form/question/question.component.tsx:1]: yup definitely, I think it should be fixed now.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx (hunk 23 above)
Comment (MohamedAliMefteh) [reply src/components/interactive-builder/modals/question/question-form/question/question.test.tsx]: I added that and I just added the test cases for other question options to check `renderingtypes` for each option.

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 24]
```diff
@@ -1,73 +1,340 @@
...
+      formField: { ...initialFormField, questionInfo: 'Initial info' },
+    });
 
-    const yesRadio = screen.getByRole('radio', { name: /Yes/i });
-    await user.click(yesRadio);
+    const questionInfoInput = screen.getByLabelText(/additional question info/i);
+    await user.clear(questionInfoInput);
+    await user.type(questionInfoInput, 'New question info');
+
+    expect((questionInfoInput as HTMLInputElement).value).toBe('New question info');
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:113]:
```suggestion
    expect(questionInfoInput).toHaveValue('New question info');
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 25]
```diff
@@ -1,73 +1,340 @@
...
+
+  it('should handle initial state with questionInfo properly', () => {
+    renderWithFormFieldProvider(<Question checkIfQuestionIdExists={checkIfQuestionIdExists} />, {
+      formField: { ...initialFormField, questionInfo: 'Initial question info' },
+    });
+
+    expect(screen.getByLabelText(/additional question info/i)).toBeInTheDocument();
+    expect((screen.getByLabelText(/additional question info/i) as HTMLInputElement).value).toBe(
+      'Initial question info',
+    );
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:325]:
```suggestion
    expect(screen.getByLabelText(/additional question info/i)).toHaveValue('Initial question info');
```

File: src/components/interactive-builder/modals/question/question-form/question/question.test.tsx [hunk 26]
```diff
@@ -1,73 +1,340 @@
...
+    expect(options).toHaveLength(renderingTypes.length);
+
+    const optionTexts = options.map((option) => option.textContent);
+
+    renderingTypes.forEach((renderType) => {
+      expect(optionTexts).toContain(renderType);
+    });
+  });
+
+  it('should handle initial state with questionInfo properly', () => {
```
Comment (NethmiRodrigo) [inline src/components/interactive-builder/modals/question/question-form/question/question.test.tsx:317]:
```suggestion
  it('should load question info from the form field object', () => {
```

**Review by NethmiRodrigo — CHANGES_REQUESTED** [review]
Thanks @MohamedAliMefteh, very close now, a few final nitpicks:

**Review by NethmiRodrigo — APPROVED** [review]
Thanks @MohamedAliMefteh!

**Comment by MohamedAliMefteh** [comment]
> Thanks @MohamedAliMefteh!

Thank you for the help!
